### PR TITLE
nspawn: set shared propagation mode for the container

### LINKED ISSUE #3860: machinectl bind does not make the new bind available to some systemd service inside the container
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> tested on systemd 230 but should be the same later
### Used distribution

> Fedora-22 in a nspawn container
### In case of bug report: Steps to reproduce the problem

Preconditions:
- Distribution installed in $HOME/distro-trees/fedora-22
- Distribution installed in $HOME/distro-trees/fedora-22/fedora-tree

Enter the container:

```
sudo systemd-nspawn -D $HOME/distro-trees/fedora-22 -b
```

In the container, prepare a service that will run in a different mount namespace:

```
# cat /usr/lib/systemd/system/test01.service 
[Service]
Type=oneshot
ExecStart=/bin/cat /etc/machine-id ; /bin/sleep 600
RootDirectory=/fedora-tree/
MountFlags=shared
# systemctl start test01.service
...
```

While test01.service is initializing or running, from the host:

```
# machinectl bind fedora-22 /tmp /fedora-tree/data
```

Check whether the "data" bind mount is really available in the container:

```
# cat /proc/1/mountinfo |grep /data
```

Check whether the "data" bind mount is really available in the test01.service:

```
# cat /proc/$(pidof sleep)/mountinfo |grep /data
```
### In case of bug report: Expected behaviour you didn't see

```
# cat /proc/1/mountinfo
851 760 253:3 /alban/distro-trees/fedora-22 / rw,relatime master:31 <shared:xxx>- ext4 /dev/mapper/fedora_rainbow-home rw,seclabel,data=ordered
1041 851 0:37 / /fedora-tree/data rw master:26 - tmpfs tmpfs rw,seclabel
# cat /proc/$(pidof sleep)/mountinfo  
1024 992 253:3 /alban/distro-trees/fedora-22/fedora-tree / rw,relatime shared:392 master:<xxx>
<line with the "data" mount>
```

In the mount namespace of the test01 service, I expect to see
- the "data" mount
- the "/" mount being enslaved to the "/" of pid 1 of the container.
### In case of bug report: Unexpected behaviour you saw

```
# cat /proc/1/mountinfo
851 760 253:3 /alban/distro-trees/fedora-22 / rw,relatime master:31 - ext4 /dev/mapper/fedora_rainbow-home rw,seclabel,data=ordered
1041 851 0:37 / /fedora-tree/data rw master:26 - tmpfs tmpfs rw,seclabel
# cat /proc/$(pidof sleep)/mountinfo  
1024 992 253:3 /alban/distro-trees/fedora-22/fedora-tree / rw,relatime shared:392 master:31 - ext4 /dev/mapper/fedora_rainbow-home rw,seclabel,data=ordered
```

In the mount namespace of the test01 service, I saw
- no "data" mount at all
- the "/" mount being enslaved to a host peer group (master:31)

/cc @alepuccetti @euank @yifan-gu @tmrts @iaguis

### PR DESCRIPTION
**Pull request by alban**
Fixes #3860
Supersedes #4152

This new version of the fix only changes systemd-nspawn (and not systemd anymore).

@poettering @tixxdz @iaguis @lucab @jonboulle

### CODE REVIEW

File: src/nspawn/nspawn.c [hunk 1]
```diff
@@ -1755,6 +1755,9 @@ static int setup_propagate(const char *root) {
         if (mount(NULL, q, NULL, MS_BIND|MS_REMOUNT|MS_RDONLY, NULL) < 0)
                 return log_error_errno(errno, "Failed to make propagation mount read-only");
 
+        if (mount(NULL, q, NULL, MS_SLAVE, NULL) < 0)
+                return log_error_errno(errno, "Failed to make propagation mount slave");
```
Comment (alban) [inline src/nspawn/nspawn.c:1761]: Add a comment:

```
/* machined will MS_MOVE into that directory, and that's only supported for non-shared mounts. */
```

File: src/nspawn/nspawn.c [hunk 2]
```diff
@@ -2990,6 +2993,13 @@ static int outer_child(
         if (mount(directory, directory, NULL, MS_BIND|MS_REC, NULL) < 0)
                 return log_error_errno(errno, "Failed to make bind mount: %m");
 
+        /* Mark everything as shared so our mounts get propagated down. This is
+         * required to make new bind mounts available in systemd services
+         * inside the containter that create a new mount namespace.
+         * See https://github.com/systemd/systemd/issues/3860 */
```
Comment (alban) [inline src/nspawn/nspawn.c]: Add a further comment:

Further submounts (such as /dev) done after this will inherit the shared propagation mode.

**Comment by alban** [comment]
Patch updated.

**Comment by iaguis** [comment]
lgtm

**Pull request merged**